crates/paimon/src/file_index/file_index_format.rs

```diff
@@ -0,0 +1,514 @@
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜  redundant length ｜redundant bytes ｜
+/// ｜-------------------------------------｜    ---------------------
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜             BODY
+/// ｜                BODY                 ｜
+/// ｜_____________________________________｜    _____________________
+///
+/// - `magic`: 8 bytes long
+/// - `version`: 4-byte integer
+/// - `head length`: 4-byte integer
+/// - `column number`: 4-byte integer
+/// - `column x`: variable-length UTF-8 string (length + bytes)
+/// - `index number`: 4-byte integer (number of index items below)
+/// - `index name x`: variable-length UTF-8 string
+/// - `start pos`: 4-byte integer
+/// - `length`: 4-byte integer
+/// - `redundant length`: 4-byte integer (for compatibility with future versions; content is zero in this version)
+/// - `redundant bytes`: variable-length bytes (for compatibility with future versions; empty in this version)
+/// - `BODY`: sequence of index data (concatenated index data for each column)
+/// ```
+#[derive(Debug)]
+pub struct FileIndexFormatWriter {
+    writer: OutputFile,
+}
+
+impl FileIndexFormatWriter {
+    pub fn new(writer: OutputFile) -> Self {
+        Self { writer }
+    }
+
+    pub async fn write_column_indexes(
+        &mut self,
+        indexes: HashMap<String, HashMap<String, Bytes>>,
+    ) -> crate::Result<()> {
+        let mut body_info: HashMap<String, HashMap<String, IndexInfo>> = HashMap::new();
+        let mut body = BytesMut::new();
+
+        for (column_name, bytes_map) in indexes.into_iter() {
+            let inner_map = body_info.entry(column_name.clone()).or_default();
+            for (index_name, data) in bytes_map {
+                let start_position = body.len() as i32;
+                if data.is_empty() {
+                    inner_map.insert(
+                        index_name,
+                        IndexInfo {
+                            start_pos: EMPTY_INDEX_FLAG,
+                            length: 0,
+                        },
+                    );
+                } else {
+                    body.extend_from_slice(&data);
+                    inner_map.insert(
+                        index_name,
+                        IndexInfo {
+                            start_pos: start_position,
+                            length: body.len() as i32 - start_position,
+                        },
+                    );
+                }
+            }
+        }
+
+        let head = self.write_head(&body_info).await?;
+        let mut combined = BytesMut::with_capacity(head.len() + body.len());
+        combined.extend_from_slice(&head);
+        combined.extend_from_slice(&body);
```

Comment:
Please write directly into `writer` to avoid extra alloc and memory usage unless you are certain that the data is small enough to be stored in memory.

Reply:
I will avoid this behavior based on the explicit "finish" method above.